src/network_service/sending_targets_cache.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::ConnectionInfo;
+use std::collections::HashMap;
+use std::net::SocketAddr;
+
+const MAX_RESENDS: u8 = 1;
+
+pub enum TargetState {
+    Sending(u8),
+    Failed(u8),
+    Sent,
+}
+
+pub struct SendingTargetsCache {
+    cache: HashMap<u64, Vec<(ConnectionInfo, TargetState)>>,
+}
+
+impl SendingTargetsCache {
+    pub fn new() -> Self {
+        Self {
+            cache: HashMap::new(),
+        }
+    }
+
+    pub fn insert_message(
+        &mut self,
+        msg_id: u64,
+        initial_targets: Vec<ConnectionInfo>,
+        dg_size: usize,
+    ) {
+        let targets = initial_targets
+            .into_iter()
+            .enumerate()
+            .map(|(idx, tgt_info)| {
+                (
+                    tgt_info,
+                    if idx < dg_size {
+                        TargetState::Sending(0)
+                    } else {
+                        TargetState::Failed(0)
+                    },
+                )
+            })
+            .collect();
+        let _ = self.cache.insert(msg_id, targets);
+    }
+
+    fn fail_target(&mut self, msg_id: u64, target: SocketAddr) {
+        for (tgt, state) in self.cache.get_mut(&msg_id).into_iter().flatten() {
```

Comment:
Is that not a `find`?

Reply:
It pretty much is, but does `find` return a mutable reference, or is there something like `find_mut`?

Reply:
Nevermind, it'll just be a `find` on `iter_mut`. I'll change this.

Reply:
Had to try it, but yes you can:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=df59978e13abb3b116d8d1bbb333697d
```
fn main() {
    let mut col = vec![1,2,3];
    col.iter_mut().find(|v| **v == 2).map(|v| *v = 5);
    println!("{:?}", col);
}
```